### PR TITLE
Implement dataset index and add PSTracks v4.2, NT v5.1

### LINKED ISSUE #181: Cluster scipy version mismatch (No module named 'scipy.interpolate._fitpack2')
**Describe the bug**
This is a follow-up on one of the issues raised in #132 and related to #35 .

Loading previously created spline pickles in `make_SoB_splines.py` fails if the spline pickles have been created with a more recent version of `scipy`.

The typical situation in which this occurs is when trials are firstly run from an up-to-date environment (such as a custom installation of python / flarestack) and then more trials are run on the cluster.

The jobs fail with:
```
  File "[...]/flarestack/flarestack/utils/make_SoB_splines.py", line 597, in load_bkg_spatial_spline
    res = Pickle.load(f)
ModuleNotFoundError: No module named 'scipy.interpolate._fitpack2'
```

**To Reproduce**
1. Install flarestack locally after preparing a `conda` environment with the prerequisites.
2. Set up an analysis and run a few trials locally.
3. Submit more trials to the cluster.

**Expected behavior**
The cluster processing should not fail.

**Additional context**
**Proposed solution**: tentatively, as long as we do not have a solution to ensure the proper environment consistency between the cluster and the user installation, we could catch the exception and trigger the recreation of the splines.

### LINKED ISSUE #182: Retrieve dataset by name instead of importing object (build a dataset index)
**Is your feature request related to a problem? Please describe.**
Aside from navigating the source code, there is no easy way (as far as I know) to see which data sets are available.
In addition, datasets must be always retrieved by explicit imports and cannot be loaded by name (see #178).

**Describe the solution you'd like**
Every created dataset could take care of indexing itself in a central location. Such index will be exposed to the user, that will be allowed to load a dataset by key or name.

This could be achieved by instantiating a dictionary inside a module under `flarestack.data`, to which every dataset will "register" (more or less like it is done do for subclasses of the core modules). 

**Describe alternatives you've considered**
Well, the current design works but it's less user friendly and provides little flexibility.

### LINKED ISSUE #207: Let `ResultsHandler` constructor fail gracefully while storing a `valid` flag
**Is your feature request related to a problem? Please describe.**
I don't like when my code is impolite and throws exceptions all around. Right now, the instantiation of `ResultsHandler` is very prone to throwing exceptions and this makes the rest of the execution of the script or notebook quite ugly.

**Describe the solution you'd like**
`ResultsHandler` should catch as many exceptions as possible and store a boolean flag `valid`, that is set to `False` if something goes wrong. Then the user can check whether `is_valid()` or not.

**Describe alternatives you've considered**
Leave it as it is, but I like my code to behave politely.

**Comments**
Work in progress as part of #180.

### LINKED ISSUE #209: High memory usage with Northern Tracks
Running analysis with Northern Tracks requires very large amounts of memory, approximately 10 GB per season. This seems to be related to the injector instances. A test job for a stacking using 2017-2021 seasons results in the following:
```
 Max vmem         = 49.945G
 Max rss          = 47.731G
```
This makes it hardly sustainable for the use on a cluster and even on a shared worker machine it could make people unhappy.

I am not sure how this can be improved, feedback is welcome!

### PR DESCRIPTION
**Pull request by mlincett**
This PR:
- implements a dataset index, with documentation, closes #182 
- adds interfaces to PSTracks v4.2 and NorthernTracks v5.1 (the two datasets can be accessed through the index);
- contrarily to previous versions, NTv5.1 uses a combo data season to fix #209 (similar to PSTracks);
- adds deprecation warning for PSTracks v4.0;
- improves `ResultsHandler` handling of errors, closes #207;
- let the DESY cluster use the user's python installation, fixes #181 (on WIPAC this is not required);
- fixes a data path handling regression introduced with one of my previous PRs.

### CODE REVIEW

**Comment by mlincett** [comment]
Requesting a review!

Please note that the commit history is a bit chaotic due to some kind of merge/rebase mistake on my part. Will make sure to squash everything on merge.

**Comment by robertdstein** [comment]
Thanks for all this @mlincett! One broad comment though: this is a lot of changes which are not really all directly coupled together. As a general rule, I think one PR per issue is optimal, and here at least I'd say this could have been more easy to review if it was several independent PRs. Reviewing smaller pieces is easier than doing it all at once.

File: flarestack/cluster/make_desy_cluster_script.py [hunk 1]
```diff
@@ -2,6 +2,10 @@
 import os
 from flarestack.shared import fs_dir, log_dir, fs_scratch_dir
 
+"""
```
Comment (robertdstein) [inline flarestack/cluster/make_desy_cluster_script.py:5]: You could actually give a warning on the terminal, with logger.warning?

File: flarestack/cluster/submitter.py [hunk 2]
```diff
@@ -437,7 +437,9 @@ def make_cluster_submission_script(self):
             "eval $(/cvmfs/icecube.opensciencegrid.org/py3-v4.1.0/setup.sh) \n"
             "export PYTHONPATH=" + DESYSubmitter.root_dir + "/ \n"
             "export FLARESTACK_SCRATCH_DIR=" + flarestack_scratch_dir + " \n"
-            "python " + fs_dir + "core/multiprocess_wrapper.py -f $1 -n $2 \n"
+            f"{sys.executable} "
```
Comment (robertdstein) [inline flarestack/cluster/submitter.py:440]: Is there a reason this is over 3 lines? The rest is one python code line per outputted line

**Comment by mlincett** [comment]
> Thanks for all this @mlincett! One broad comment though: this is a lot of changes which are not really all directly coupled together. As a general rule, I think one PR per issue is optimal, and here at least I'd say this could have been more easy to review if it was several independent PRs. Reviewing smaller pieces is easier than doing it all at once.

You are right, and I was indeed a bit worried about this. Somehow it came natural to implement one feature/fix to help with testing another. But now that everything is there I am fine with splitting this into three/four separate PRs.

File: flarestack/data/icecube/northern_tracks/nt_v005_p01.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+from flarestack.data.icecube.ic_season import IceCubeDataset, get_dataset_dir
+from flarestack.data.icecube.northern_tracks import (
+    NTSeasonNewStyle,
+    get_diffuse_binning,
+)
+
+icecube_dataset_dir = get_dataset_dir()
```
Comment (robertdstein) [inline flarestack/data/icecube/northern_tracks/nt_v005_p01.py:7]: Is this the way the code needs to run now? i.e you need imports at the top, then `get_dataset_dir()`, then `import dataset_index`

The pythonic way is usually for all imports to go together at the top

File: flarestack/data/icecube/northern_tracks/nt_v005_p01.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+from flarestack.data.dataset_index import dataset_index
+
+nt_data_dir = icecube_dataset_dir + "northern_tracks/version-005-p01/"
```
Comment (robertdstein) [inline flarestack/data/icecube/northern_tracks/nt_v005_p01.py:11]: What type is icecube_dataset_dir? I think do an os.path.join or Pathlib equivalent would be more robust

File: flarestack/data/icecube/ps_tracks/ps_v004_p02.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+sample_name = "ps_tracks_v004_p02"
+
+dataset_path = "ps_tracks/version-004-p02"
+
+data_path = icecube_dataset_dir / dataset_path
+grl_path = data_path / "GRL"
+
+# the path handling is a bit hacky down the line, one day this will be fixed
+ps_data_dir = str(data_path) + "/"
```
Comment (robertdstein) [inline flarestack/data/icecube/ps_tracks/ps_v004_p02.py:25]: Just use Pathlib here?

**Comment by robertdstein** [comment]
> > Thanks for all this @mlincett! One broad comment though: this is a lot of changes which are not really all directly coupled together. As a general rule, I think one PR per issue is optimal, and here at least I'd say this could have been more easy to review if it was several independent PRs. Reviewing smaller pieces is easier than doing it all at once.
> 
> You are right, and I was indeed a bit worried about this. Somehow it came natural to implement one feature/fix to help with testing another. But now that everything is there I am fine with splitting this into three/four separate PRs.

I think it's too much work to split this now, but just something to bear in mind for next time

File: flarestack/cluster/submitter.py (hunk 2 above)
Comment (mlincett) [reply flarestack/cluster/submitter.py:440]: This is what `black` does.

File: flarestack/data/icecube/northern_tracks/nt_v005_p01.py (hunk 3 above)
Comment (mlincett) [reply flarestack/data/icecube/northern_tracks/nt_v005_p01.py:7]: This was unintended, indeed.

File: flarestack/data/icecube/northern_tracks/nt_v005_p01.py (hunk 4 above)
Comment (mlincett) [reply flarestack/data/icecube/northern_tracks/nt_v005_p01.py:11]: I had just copy pasted the NTv5.0 file but I agree. Gonna move to Pathlib and test it.

**Comment by mlincett** [comment]
> I think it's too much work to split this now, but just something to bear in mind for next time

Uhm, I still think it will save us a lot of headaches and make the review quicker. Will split this in five different PR!

**Pull request closed without merging**